### PR TITLE
[Bugfix] Reject heterogeneous vocab with local argmax reduction

### LINKED ISSUE #51561: [Bug]: use_heterogeneous_vocab (TLI) silently bypassed when use_local_argmax_reduction is enabled
### Your current environment

Static analysis of `main` (`llm_base_proposer.py` @ `83ad767ee`; `vocab_mapping.py` unchanged since #38174 merged). Environment-independent flag-handling defect; reproduces on any TP size, including TP=1. I have not executed the offending configuration — the reachability argument below rests on which classes mix in `LocalArgmaxMixin`.

### 🐛 Describe the bug

`use_local_argmax_reduction` and `use_heterogeneous_vocab` (TLI, #38174) are independent opt-in flags with no mutual-exclusion validator. With both enabled, the local-argmax branch in `_greedy_sample` (`vllm/v1/spec_decode/llm_base_proposer.py:431`) returns early, skipping both the intersection mask and the draft→target ID mapping:

```python
def _greedy_sample(self, hidden_states: torch.Tensor) -> torch.Tensor:
    """Greedy-sample draft tokens from hidden states."""
    if self.use_local_argmax_reduction:
        return self.model.get_top_tokens(hidden_states)   # <-- no mask, no draft->target mapping
    if self.use_heterogeneous_vocab:
        logits = self.model.compute_logits(hidden_states)
        logits = self.vocab_mapping.constrain_draft_logits(logits)
        draft_token_ids = logits.argmax(dim=-1)
        return self.vocab_mapping.map_draft_to_target_ids(draft_token_ids)
```

`LocalArgmaxMixin.get_top_tokens` (`vllm/model_executor/models/interfaces.py:1459`) returns **draft-vocabulary** indices. Its optional `draft_id_to_target_id` remap is an EAGLE/MTP offset table that neither drafter class below defines, so no translation happens. Meanwhile the input side still maps target→draft (`llm_base_proposer.py:694` in `propose`, `:837`/`:840` in `set_inputs_first_pass`), so the two directions disagree.

**Reachable with a documented config.** `LocalArgmaxMixin` is mixed into standard drafter-capable classes — `Qwen3ForCausalLM` (`qwen3.py:271`) and `LlamaForCausalLM` (`llama.py:447`) — covering the TLI docs' own pairing (`Qwen/Qwen3-8B` target + `HuggingFaceTB/SmolLM2-135M-Instruct` draft, a `LlamaForCausalLM`). `LogitsProcessor.get_top_tokens` has an explicit `if tp_size == 1: return global_indices` (`logits_processor.py:191`), so TP=1 is affected too. Both flags default to `False`, so this is latent rather than a live regression.

Note that the V2 runner's equivalent path already validates a *different* pairwise conflict for this same flag — `Speculator._validate_local_argmax_reduction` (`vllm/v1/worker/gpu/spec_decode/speculator.py:290`) rejects `use_local_argmax_reduction` + `draft_sample_method='probabilistic'` — so upfront pairwise validation is the established pattern here. The V1 check at `llm_base_proposer.py:1616` only verifies that `get_top_tokens` exists.

### Impact

**Not a losslessness violation.** `use_heterogeneous_vocab` is already constrained to `draft_sample_method='greedy'`, so `draft_probs is None` and the ratio test degenerates to `min(1, p(x)/1)`. A wrong draft ID can only cause a spurious *rejection*; the emitted token remains exactly target-distributed. The impact is:

1. **Silent acceptance-rate collapse.** `rejection_greedy_sample_kernel` tests `draft_token_id != target_argmax_id`, now comparing a draft-space ID against a target-space ID. These coincide only by accident, so acceptance tends toward zero and speculation becomes pure overhead — slower than running without speculative decoding, with no warning logged.
2. **Out-of-bounds read on `target_probs` for `temperature > 0` requests when `draft_vocab_size > target_vocab_size`.** In `rejection_random_sample_kernel`, `draft_token_id` indexes `target_probs` directly and unmasked:

   ```python
   target_prob = tl.load(
       target_probs_ptr + (start_idx + pos) * vocab_size + draft_token_id
   )
   ```

   Nothing bounds an unmapped draft ID by the target vocab size: `rejection_sample` asserts only shape/contiguity, never `draft_token_ids.max() < vocab_size`, and no clamp exists on the proposer side. This requires a non-greedy request — all-greedy batches return early at the `if sampling_metadata.all_greedy: return output_token_ids` guard and never reach this kernel. The documented pairing (151936 target vs 49152 draft) stays in range, but e.g. `Qwen/Qwen3-0.6B` drafting (151936, has the mixin) for a `TinyLlama-1.1B` target (32000) does not.

### Suggested fix

Local argmax cannot apply the intersection mask without gathering full logits, so the two features are incompatible as written. Reject upfront in `SpeculativeConfig._verify_args`, beside the existing `draft_sample_method` guard (`vllm/config/speculative.py:1386`):

```python
if self.use_heterogeneous_vocab and self.use_local_argmax_reduction:
    raise ValueError(
        "use_heterogeneous_vocab is incompatible with "
        "use_local_argmax_reduction: local argmax cannot apply the "
        "vocabulary-intersection mask or the draft->target ID mapping."
    )
```

### Notes

Drafted with AI assistance (Claude) and reviewed by me before filing; flagging it since there is no code diff for a reviewer to check the claims against. Happy to be corrected on any file:line citation. Disclosure of interest: I'm a co-author of the TLI paper referenced by #38174, with no involvement in the vLLM implementation.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the [Discussions](https://github.com/vllm-project/vllm/discussions)

### PR DESCRIPTION
**Pull request by 030611**
## Purpose

Fixes #51561.

The local-argmax path returns before the heterogeneous-vocabulary mask and draft-to-target mapping run. Rejecting the two flags together in `SpeculativeConfig` prevents this unsupported combination from silently reaching the sampler. Other flag combinations are unchanged.

This is separate from #42775 (optional vocabulary padding) and #54780 (tokenizer mode/revision handling). The issue and related PRs were rechecked on September 7.

## Validation

The original guard and regression test have been refreshed onto `f2d45f26`:

- Without the guard, the committed regression test fails with `DID NOT RAISE ValueError`.
- With the guard, the regression and two neighboring configuration tests pass. Three additional checks confirm the supported flag combinations still work: **6 isolated checks passed**.
- Ruff 0.14.0 check/format, whitespace checks, and six relevant repository check scripts pass: SPDX headers, lazy imports, forbidden imports, CUDA API usage, config validation, and boolean context managers.

These Windows CPU checks load the current production configuration module and the unchanged test functions, with transitive dependencies from the previous checkout. They are not a full current-main installation or full-file test run. Ruff explicitly classifies `vllm` as first-party because the checkout is sparse. Full pre-commit dependency initialization did not complete.

In a standard vLLM test environment, the committed regression is:

```bash
pytest tests/test_config.py -k heterogeneous_vocab_rejects_local_argmax_reduction
```

No GPU or model evaluation was run: this adds validation for an unsupported configuration and does not change a valid serving path.

Codex assisted with the patch refresh and checks; I reviewed the patch.

### CODE REVIEW

File: vllm/config/speculative.py [hunk 1]
```diff
@@ -1777,6 +1777,13 @@ def _verify_args(self) -> Self:
                 "omit it."
             )
 
+        if self.use_heterogeneous_vocab and self.use_local_argmax_reduction:
+            raise ValueError(
```
Comment (tomasruizt) [inline vllm/config/speculative.py:1782]: For future context, you could add a comment with the issue describing the bug https://github.com/vllm-project/vllm/issues/51561

File: tests/test_config.py [hunk 2]
```diff
@@ -2532,6 +2532,19 @@ def test_qwen3_omni_dspark_allows_smaller_input_vocabulary():
...
+def test_heterogeneous_vocab_rejects_local_argmax_reduction():
+    speculative_config = SpeculativeConfig(
+        method="ngram",
+        num_speculative_tokens=1,
+    )
+    speculative_config.method = "draft_model"
+    speculative_config.use_heterogeneous_vocab = True
+    speculative_config.use_local_argmax_reduction = True
+
+    with pytest.raises(ValueError, match="cannot be used together"):
```
Comment (tomasruizt) [inline tests/test_config.py]: The match string could be more informative. E.g. 
"use_heterogeneous_vocab and use_local_argmax_reduction cannot be used together"

File: vllm/config/speculative.py (hunk 1 above)
Comment (030611) [reply vllm/config/speculative.py:1782]: Added the issue link next to the guard in 94875b1f. Thanks!

File: tests/test_config.py (hunk 2 above)
Comment (030611) [reply tests/test_config.py]: Updated the match to include both option names in 94875b1f. The focused CPU checks pass.
